Assets/World/Buildings/Agricultural/Farm2D/farm2D.gd

```diff
@@ -1,42 +1,38 @@
-extends Node2D
+extends ProductionBuilding2D
 
-@export_category("Farm")
-@export var game_name: String = "Farm"
-@export var production_time: int = 10
-@export var max_storage_capacity: int = 10
-@export var output_product: String
-@export var needs_intake_poduct: bool = false
-
-@onready var prod_timer: Timer = get_node("ProdTimer")
-#@onready var carrier: Worker = get_node("Person")
-
-var number_of_output_products = 0
-var number_of_intake_products = 0
+class_name Farm2D
 
+#@export_category("Farm2D")
+@export var production_time: int = 10
 
 
 
 func _ready():
-  var path = self.get_parent().get_path_to_dest(self.position, Vector2(0, 0))
-  if path != null:
-    self.carrier.path = path
+  #var path = self.get_parent().get_path_to_dest(self.position, Vector2(0, 0))
+  #if path != null:
+    #self.carrier.path = path
+  self.set_up()
   prod_timer.start(production_time)
 
 
 
-func _on_timer_timeout():
-  if needs_intake_poduct:
-    if number_of_intake_products > 0:
-      number_of_intake_products -= 1
+func _on_prod_timer_timeout():
+  if self.needs_intake_product:
+    if self.number_of_intake_products > 0:
+      self.number_of_intake_products -= 1
     else:
       return
 
-  if number_of_output_products < 10:
-    number_of_output_products += 1
-    
+  if self.number_of_output_products < 10:
+    self.number_of_output_products += 1
+
+  var objects_to_carry: Dictionary = {"to_warehouse": ["", 0], "back": null}
+  if not carrier.is_moving or len(closest_warehouse_path) > 0:
+    var amount_to_load = min(number_of_output_products - carrier.objects_carring.get("to_warehouse")[1], carrier.max_carry_limit)
+    objects_to_carry["to_warehouse"] = [output_product, amount_to_load + objects_to_carry["to_warehouse"][1]]
+
+    if input_product == "":  objects_to_carry["back"] = null
+    else:  objects_to_carry["back"] = min(max_storage_capacity - number_of_intake_products, carrier.max_carry_limit)
 
-  if not self.carrier.is_moving:
-    self.carrier.objects_carring["there"] = [output_product, min(number_of_output_products, self.carrier.max_carry_limit)]
-    self.carrier.objects_carring["back"] = null
-    number_of_output_products -= self.carrier.count_of_objects
-    self.carrier.move()
+    carry_resources.emit(objects_to_carry)
+    resource_produced.emit(output_product, 1)
```

Comment:
1) unify two events, should be one, and the recipients decide what parameters to use
2) worker should request the amount of resources to bring back when at the warehouse, not at the production unit